File: my-go-backend/internal/models/widget.go
package models

import (
	"fmt"
	"time"

	"gorm.io/datatypes"
)

// Widget مدل ابزارک - هر ابزارک یک نمونه منحصر به فرد با محتوای اختصاصی است
type Widget struct {
	ID           uint           `json:"id" gorm:"primaryKey"`                                  // کلید اصلی
	Code         string         `json:"code" gorm:"type:varchar(100);uniqueIndex;not null"`    // کد یونیک برای استفاده در صفحات
	Title        string         `json:"title" gorm:"type:varchar(255);not null"`               // عنوان ابزارک
	Description  string         `json:"description" gorm:"type:text"`                          // توضیحات ابزارک
	Type         string         `json:"type" gorm:"type:varchar(50);not null;index"`           // نوع ابزارک (slider, banner, product-grid, ...)
	Status       string         `json:"status" gorm:"type:varchar(20);default:'active';index"` // وضعیت (active, inactive, draft)
	SortOrder    int            `json:"sort_order" gorm:"column:sort_order;default:1;index"`   // ترتیب نمایش
	Image        string         `json:"image" gorm:"type:varchar(500)"`                        // تصویر پیش‌نمایش ابزارک
	Link         string         `json:"link" gorm:"type:varchar(500)"`                         // لینک ابزارک (در صورت نیاز)
	Page         string         `json:"page" gorm:"type:varchar(50);default:'home';index"`     // صفحه‌ای که ابزارک در آن نمایش داده می‌شود
	ShowOnMobile bool           `json:"show_on_mobile" gorm:"default:true"`                    // نمایش در موبایل
	Config       datatypes.JSON `json:"config" gorm:"type:jsonb"`                              // تنظیمات و محتوای اختصاصی ابزارک (JSON)
	CreatedAt    time.Time      `json:"created_at"`
	UpdatedAt    time.Time      `json:"updated_at"`
}

// WidgetConfig ساختار پایه برای تنظیمات ابزارک
type WidgetConfig struct {
	// تنظیمات عمومی
	BackgroundColor string `json:"background_color,omitempty"`
	Padding         string `json:"padding,omitempty"`
	Margin          string `json:"margin,omitempty"`
	CustomClass     string `json:"custom_class,omitempty"`
}

// SliderConfig تنظیمات اسلایدر
type SliderConfig struct {
	WidgetConfig
	SliderCount    int          `json:"slider_count"`
	WideBg         bool         `json:"wide_bg"`
	BgColor        string       `json:"bg_color"`
	BannerPosition string       `json:"banner_position"` // left, right
	DisplayOrder   string       `json:"display_order"`   // asc, desc
	Height         int          `json:"height"`
	Slides         []SlideItem  `json:"slides"`
	SideBanners    []BannerItem `json:"side_banners"`
}

// BannerConfig تنظیمات بنر
type BannerConfig struct {
	WidgetConfig
	Layout  string       `json:"layout"` // single, double, triple, quad
	Banners []BannerItem `json:"banners"`
}

// ProductConfig تنظیمات محصولات
type ProductConfig struct {
	WidgetConfig
	CategoryID   uint   `json:"category_id,omitempty"`
	ProductCount int    `json:"product_count"`
	Layout       string `json:"layout"` // grid, list, carousel
	ShowPrice    bool   `json:"show_price"`
	ShowRating   bool   `json:"show_rating"`
	ShowDiscount bool   `json:"show_discount"`
}

// SlideItem آیتم اسلایدر
type SlideItem struct {
	ID          uint   `json:"id,omitempty"`
	Title       string `json:"title"`
	Description string `json:"description,omitempty"`
	Image       string `json:"image"`
	Link        string `json:"link,omitempty"`
	Order       int    `json:"order"`
	Status      string `json:"status"` // active, inactive
}

// BannerItem آیتم بنر
type BannerItem struct {
	ID          uint   `json:"id,omitempty"`
	Title       string `json:"title"`
	Description string `json:"description,omitempty"`
	Image       string `json:"image"`
	Link        string `json:"link,omitempty"`
	Order       int    `json:"order"`
	Status      string `json:"status"` // active, inactive
}

// ProductCarouselConfig تنظیمات کاروسل محصولات
type ProductCarouselConfig struct {
	WidgetConfig
	Title         string `json:"title"`
	Subtitle      string `json:"subtitle"`
	CategoryID    *uint  `json:"category_id,omitempty"`
	ProductCount  int    `json:"product_count"`
	SlidesPerView int    `json:"slides_per_view"`
	ShowPrice     bool   `json:"show_price"`
	ShowRating    bool   `json:"show_rating"`
	ShowDiscount  bool   `json:"show_discount"`
	Autoplay      bool   `json:"autoplay"`
	AutoplayDelay int    `json:"autoplay_delay"`
	WideBg        bool   `json:"wide_bg"`
	BgColor       string `json:"bg_color"`
}

// CategoryConfig تنظیمات دسته‌بندی
type CategoryConfig struct {
	WidgetConfig
	Categories       []CategoryItem `json:"categories"`
	Layout           string         `json:"layout"` // grid, list
	ShowProductCount bool           `json:"show_product_count"`
	Columns          int            `json:"columns"` // تعداد ستون‌ها
	ShowTitle        bool           `json:"show_title"`
	Title            string         `json:"title"`
	TitleColor       string         `json:"title_color"`
	CardStyle        string         `json:"card_style"`     // default, rounded, shadow
	ImageSize        string         `json:"image_size"`     // small, medium, large
	TextAlignment    string         `json:"text_alignment"` // left, center, right
}

// CategoryItem آیتم دسته‌بندی
type CategoryItem struct {
	ID           uint   `json:"id"`
	Title        string `json:"title"`
	Description  string `json:"description,omitempty"`
	Image        string `json:"image"`
	Link         string `json:"link"`
	ProductCount int    `json:"product_count,omitempty"`
}

// تابع تولید کد یونیک
func (w *Widget) GenerateUniqueCode() string {
	typeMap := map[string]string{
		"main-slider-side-banner": "SLIDER_MAIN",
		"single-slider-side":      "SLIDER_SINGLE",
		"full-slider":             "SLIDER_FULL",

		"products-box-bg": "PRODUCT_BG",
		"full-banner":     "BANNER_FULL",

		"double-banner":    "BANNER_DOUBLE",
		"triple-banner":    "BANNER_TRIPLE",
		"quad-banner":      "BANNER_QUAD",
		"penta-banner":     "BANNER_PENTA",
		"brands-slider":    "BRANDS_SLIDER",
		"services-slider":  "SERVICES_SLIDER",
		"category-box":     "CATEGORY_BOX",
		"product-carousel": "PRODUCT_CAROUSEL",
		"blog-posts":       "BLOG_POSTS",
	}

	prefix := typeMap[w.Type]
	if prefix == "" {
		prefix = "WIDGET"
	}

	// در اینجا باید شماره یونیک تولید کنید (می‌تواند از database count استفاده کند)
	return prefix + "_001" // برای الان ثابت، بعداً dynamic می‌شود
}

// تابع validate برای مدل
func (w *Widget) Validate() error {
	if w.Title == "" {
		return fmt.Errorf("عنوان ابزارک الزامی است")
	}
	if w.Type == "" {
		return fmt.Errorf("نوع ابزارک الزامی است")
	}
	validTypes := []string{
		"main-slider-side-banner", "single-slider-side", "full-slider",
		"products-box-bg", "full-banner",
		"double-banner", "triple-banner", "quad-banner", "penta-banner",
		"brands-slider", "services-slider", "category-box", "blog-posts",
		"product-carousel",
	}

	isValidType := false
	for _, validType := range validTypes {
		if w.Type == validType {
			isValidType = true
			break
		}
	}

	if !isValidType {
		return fmt.Errorf("نوع ابزارک نامعتبر است")
	}

	return nil
}

File: app/types/widget.ts
// انواع ابزارک‌ها
export type WidgetType =
  | 'main-slider-side-banner'
  | 'single-slider-side'
  | 'full-slider'

  | 'full-banner'

  | 'double-banner'
  | 'triple-banner'
  | 'quad-banner'
  | 'penta-banner'
  | 'brands-slider'
  | 'services-slider'
  | 'category-box'
  | 'blog-posts'
  | 'product-carousel'
  | 'mobile-header';

// وضعیت ابزارک
export type WidgetStatus = 'active' | 'inactive' | 'draft';

// صفحات قابل استفاده
export type WidgetPage = 'home' | 'category' | 'product' | 'about' | 'contact';

// رابط اصلی ابزارک
export interface Widget {
  id: number;
  code: string;
  title: string;
  description?: string;
  type: WidgetType;
  status: WidgetStatus;
  order: number;
  image?: string;
  link?: string;
  page: WidgetPage;
  show_on_mobile: boolean;
  config: WidgetConfig;
  created_at: string;
  updated_at: string;
}

// تنظیمات پایه ابزارک
export interface WidgetConfig {
  mobile_height?: number | 'auto'; // ارتفاع اسلایدر موبایل
  show_in_mobile?: boolean; // نمایش در موبایل
  padding?: string;
  margin?: string;
  custom_class?: string;
}

// تنظیمات اسلایدر
export interface SliderConfig extends WidgetConfig {
  wide_bg: boolean;
  bg_color: string;
  bg_enabled: boolean;
  banner_position?: 'left' | 'right';
  slider_banner_ratio?: string; // نسبت اسلایدر و بنر (مثل "7-5", "6-6", "5-7")
  height: number;
  padding_top: number;
  padding_bottom: number;
  margin_left: number;
  margin_right: number;
  autoplay_delay: number;
  autoplay_enabled: boolean;
  navigation_enabled: boolean;
  slides: SlideItem[];
  side_banners?: BannerItem[];
  mobile_slides?: SlideItem[];
  mobile_side_banners?: BannerItem[];
  // تنظیمات جدید اسلایدر تکی
  slider_count?: number;
  display_order?: 'asc' | 'desc';
  show_navigation?: boolean;
  show_pagination?: boolean;
  show_title?: boolean;
  show_description?: boolean;
  transition_duration?: number;
  navigation_style?: 'default' | 'minimal' | 'bold';
  pagination_style?: 'default' | 'small' | 'large';
  title_style?: 'default' | 'custom';
  loop_enabled?: boolean;
  pause_on_hover?: boolean;
  // تنظیمات ظاهری
  border_radius?: number;
  border_width?: number;
  border_color?: string;
  shadow_enabled?: boolean;
  shadow_style?: 'default' | 'large';
  navigation_size?: number;
  pagination_color?: string;
  title_background_opacity?: number;
  title_text_color?: string;
  title_font_size?: number;
  // Mobile settings - تنظیمات کامل موبایل
  mobile_height?: number | 'auto';
  mobile_padding_top?: number;
  mobile_padding_bottom?: number;
  mobile_margin_left?: number;
  mobile_margin_right?: number;
  mobile_show_slider?: boolean;
  mobile_bg_enabled?: boolean;
  mobile_bg_color?: string;
  mobile_wide_bg?: boolean;
  // Additional mobile settings for slider
  mobile_autoplay_enabled?: boolean;
  mobile_autoplay_delay?: number;
  mobile_loop_enabled?: boolean;
  mobile_show_navigation?: boolean;
  mobile_show_pagination?: boolean;
  mobile_show_title?: boolean;
  mobile_show_description?: boolean;
  mobile_vertical_display?: boolean;
  mobile_slider_width?: number | 'auto';
  mobile_banner_width?: number | 'auto';
  mobile_banner_height?: number | 'auto';
  mobile_banner_position?: 'top' | 'bottom';
  // Easy load feature
  easy_load_enabled?: boolean;
  // Desktop slider width settings
  slider_width?: number;
  center_width?: number;
  banner_width?: number;
  // نسبت‌های بنر برای بنرهای چندتایی
  banner1_ratio?: number;
  banner2_ratio?: number;
  banner3_ratio?: number;
  banner4_ratio?: number;
  banner5_ratio?: number;
}

// تنظیمات بنر
export interface BannerConfig extends WidgetConfig {
  layout?: 'single' | 'double' | 'triple' | 'quad';
  banners: BannerItem[];
  mobile_banners?: BannerItem[];  // آرایه جداگانه برای بنرهای موبایل
  // فیلدهای اضافی برای تنظیمات پیشرفته
  wide_bg?: boolean;
  bg_color?: string;
  bg_enabled?: boolean;
  height?: number;
  banner_width?: number;
  center_width?: number;
  border_radius?: number;
  gap?: number;
  display_order?: 'asc' | 'desc';
  layout_type?: 'horizontal' | 'vertical' | 'grid';
  // فیلدهای پایه که از WidgetConfig می‌آیند
  background_color?: string;
  padding?: string;
  margin?: string;
  custom_class?: string;
  // پدینگ و مارجین
  padding_top?: number;
  padding_bottom?: number;
  margin_left?: number;
  margin_right?: number;
  // تنظیمات نمایش
  show_title?: boolean;
  show_description?: boolean;
  // تنظیمات ظاهری
  shadow_enabled?: boolean;
  shadow_style?: 'default' | 'large';
  // تنظیمات موبایل
  mobile_vertical_display?: boolean;
  mobile_height?: number;
  mobile_banner_width?: number; // عرض بنر موبایل
  mobile_slider_width?: number; // عرض اسلایدر در موبایل
  mobile_padding_top?: number;
  mobile_padding_bottom?: number;
  mobile_margin_left?: number;
  mobile_margin_right?: number;
  // تنظیمات عکس موبایل
  mobile_image_mode?: 'auto' | 'separate';
  mobile_crop_width?: number; // عرض پیش‌فرض برای برش موبایل
  mobile_crop_height?: number; // ارتفاع پیش‌فرض برای برش موبایل

  mobile_cropped_image?: string; // URL عکس برش داده شده برای موبایل
  mobile_image?: string;
  // تنظیمات لیزی لود
  easy_load_enabled?: boolean;
  // نسبت‌های بنر برای بنرهای چندتایی
  banner1_ratio?: number;
  banner2_ratio?: number;
  banner3_ratio?: number;
  banner4_ratio?: number;
  banner5_ratio?: number;
}

// تنظیمات بنر تک تمام عرض
export interface FullBannerConfig extends WidgetConfig {
  title?: string;
  image?: string;
  mobile_image?: string;
  link?: string;
  openInNewTab?: boolean;
  banners?: BannerItem[];
  mobile_banners?: BannerItem[];  // آرایه جداگانه برای بنرهای موبایل
}

// تنظیمات بنر دو تایی
export interface DoubleBannerConfig extends WidgetConfig {
  banners: BannerItem[];
  mobile_banners?: BannerItem[];  // آرایه جداگانه برای بنرهای موبایل
}

// تنظیمات بنر سه تایی
export interface TripleBannerConfig extends WidgetConfig {
  banners: BannerItem[];
}

// تنظیمات بنر چهار تایی
export interface QuadBannerConfig extends WidgetConfig {
  banners: BannerItem[];
}

// تنظیمات محصولات
export interface ProductConfig extends WidgetConfig {
  category_id?: number;
  product_count: number;
  layout: 'grid' | 'list' | 'carousel';
  show_price: boolean;
  show_rating: boolean;
  show_discount: boolean;
}

// تنظیمات کاروسل محصولات
export interface ProductCarouselConfig extends WidgetConfig {
  title?: string;
  subtitle?: string;
  categoryId?: number | null;
  productCount: number;
  slidesPerView: number;
  showPrice: boolean;
  showRating: boolean;
  showDiscount: boolean;
  autoplay: boolean;
  autoplayDelay: number;
  wide_bg: boolean;
  bg_color: string;
  background_color?: string; // نگه‌داری برای سازگاری، اما توصیه به استفاده از bg_color
  padding?: string;
  margin?: string;
  showNavigation?: boolean;
  showIndicators?: boolean;
  showControls?: boolean;
  navigationStyle?: 'default' | 'minimal' | 'bold';
  navigationSize?: number;
  indicatorStyle?: 'default' | 'small' | 'large';
  indicatorColor?: string;
}

// تنظیمات دسته‌بندی
export interface CategoryConfig extends WidgetConfig {
  categories: CategoryItem[];
  layout: 'grid' | 'list';
  vertical_layout: boolean;
  show_product_count: boolean;
  columns: number;
  mobile_columns: number;
  box_width: 'full' | 'center';
  show_title: boolean;
  title: string;
  title_color: string;
  title_alignment?: 'left' | 'center' | 'right';
  card_style: 'default' | 'rounded' | 'shadow';
  image_size: 'small' | 'medium' | 'large';
  border_radius: 'none' | 'small' | 'medium' | 'large';
  text_alignment: 'left' | 'center' | 'right';
  background_color: string;
  show_background: boolean;
  full_width_background: boolean;
  show_border: boolean;
  border_color: string;
  border_width: 'thin' | 'medium' | 'thick';
  category_source: 'best_selling' | 'user_search' | 'user_interests' | 'manual';
  padding_top?: string;
  padding_bottom?: string;
}

// آیتم اسلایدر
export interface SlideItem {
  id?: number;
  title: string;
  description?: string;
  image: string;
  mobile_image?: string; // عکس جداگانه برای موبایل
  link?: string;
  openInNewTab?: boolean; // باز کردن لینک در صفحه جدید
  order: number;
  status: 'active' | 'inactive';
  showTitle?: boolean; // کنترل نمایش عنوان در اسلایدر
}

// آیتم بنر
export interface BannerItem {
  id?: number;
  title: string;
  description?: string;
  image: string;
  mobile_image?: string; // عکس جداگانه برای موبایل
  link?: string;
  openInNewTab?: boolean; // باز کردن لینک در صفحه جدید
  order: number;
  status: 'active' | 'inactive';
  showTitle?: boolean; // کنترل نمایش عنوان در بنر
}

// آیتم دسته‌بندی
export interface CategoryItem {
  id: number;
  title: string;
  description?: string;
  image: string;
  link: string;
  product_count?: number;
  // فیلدهای اضافی برای فرم ویرایش
  searchTerm?: string;
  showDropdown?: boolean;
  selectedCategory?: Record<string, unknown>;
}

// Request/Response types
export interface CreateWidgetRequest {
  title: string;
  description?: string;
  type: WidgetType;
  status?: WidgetStatus;
  page?: WidgetPage;
  show_on_mobile?: boolean;
  config?: WidgetConfig;
}

export interface UpdateWidgetRequest {
  title?: string;
  description?: string;
  type?: WidgetType;
  status?: WidgetStatus;
  order?: number;
  image?: string;
  link?: string;
  page?: WidgetPage;
  show_on_mobile?: boolean;
  config?: WidgetConfig;
}

export interface UpdateWidgetOrderRequest {
  items: Array<{
    id: number;
    order: number;
  }>;
}

export interface WidgetResponse {
  message: string;
  data: Widget;
}

export interface WidgetListResponse {
  message: string;
  data: Widget[];
}

// Form types for admin
export interface WidgetFormData {
  title: string;
  description: string;
  type: WidgetType;
  status: WidgetStatus;
  page: WidgetPage;
  config: WidgetConfig;
}

// Widget type labels (for UI)
export const WIDGET_TYPE_LABELS: Record<WidgetType, string> = {
  'main-slider-side-banner': 'اسلایدر اصلی و بنر کناری',
  'single-slider-side': 'اسلایدر تکی (بدون سایدبار)',
  'full-slider': 'اسلایدر تمام عرض',
  'full-banner': 'بنر تکی',
  'double-banner': 'بنر دوتایی',
  'triple-banner': 'بنر سه‌تایی',
  'quad-banner': 'بنر 4 تایی',
  'penta-banner': 'بنر پنج‌تایی',
  'brands-slider': 'اسلایدر برند ها',
  'services-slider': 'اسلایدر خدمات',
  'category-box': 'دسته بندی محصولات',
  'blog-posts': 'نوشته های وبلاگ',
  'product-carousel': 'کاروسل محصولات',
  'mobile-header': 'هدر موبایل',
};

// Widget status labels
export const WIDGET_STATUS_LABELS: Record<WidgetStatus, string> = {
  'active': 'فعال',
  'inactive': 'غیرفعال',
  'draft': 'پیش‌نویس',
};

// Widget page labels
export const WIDGET_PAGE_LABELS: Record<WidgetPage, string> = {
  'home': 'صفحه اصلی',
  'category': 'صفحه دسته‌بندی',
  'product': 'صفحه محصول',
  'about': 'درباره ما',
  'contact': 'تماس با ما',
};

// Validation helpers
export const validateWidget = (widget: Partial<Widget>): string[] => {
  const errors: string[] = [];

  if (!widget.title) {
    errors.push('عنوان ابزارک الزامی است');
  }

  if (!widget.type) {
    errors.push('نوع ابزارک الزامی است');
  }

  if (widget.type && !Object.keys(WIDGET_TYPE_LABELS).includes(widget.type)) {
    errors.push('نوع ابزارک نامعتبر است');
  }

  return errors;
};

// Helper functions
export const getWidgetTypeLabel = (type: WidgetType): string => {
  return WIDGET_TYPE_LABELS[type] || type;
};

export const getWidgetStatusLabel = (status: WidgetStatus): string => {
  return WIDGET_STATUS_LABELS[status] || status;
};

export const getWidgetPageLabel = (page: WidgetPage): string => {
  return WIDGET_PAGE_LABELS[page] || page;
};

export const generateWidgetCode = (type: WidgetType, count: number): string => {
  const typeMap: Record<WidgetType, string> = {
    'main-slider-side-banner': 'SLIDER_MAIN',
    'single-slider-side': 'SLIDER_SINGLE',
    'full-slider': 'SLIDER_FULL',
    'full-banner': 'BANNER_FULL',
    'double-banner': 'BANNER_DOUBLE',
    'triple-banner': 'BANNER_TRIPLE',
    'quad-banner': 'BANNER_QUAD',
    'penta-banner': 'BANNER_PENTA',
    'brands-slider': 'BRANDS_SLIDER',
    'services-slider': 'SERVICES_SLIDER',
    'category-box': 'CATEGORY_BOX',
    'blog-posts': 'BLOG_POSTS',
    'product-carousel': 'PRODUCT_CAROUSEL',
    'mobile-header': 'MOBILE_HEADER',
  };

  const prefix = typeMap[type] || 'WIDGET';
  return `${prefix}_${String(count).padStart(3, '0')}`;
};

// interface برای محصولات
export interface Product {
  id: number;
  name: string;
  image?: string;
  price: number;
  original_price?: number;
  discount_percentage?: number;
  rating?: number;
  rating_count?: number;
  sku?: string | number;
  slug?: string | null;
  seo_title?: string | null;
  features?: string[];
  stock_quantity?: number;
  in_stock?: boolean;
  call_for_price?: boolean;
  created_at?: string;
  updated_at?: string;
} 